File: archive/pulumi-py/Pr7669/lib/lambda_stack.py
"""
lambda_stack.py

Lambda functions for payment processing and session management.
"""

import pulumi
import pulumi_aws as aws
from pulumi import ResourceOptions, Output
from typing import Optional, List
import json


class LambdaStack(pulumi.ComponentResource):
    """Lambda functions for API backend."""

    def __init__(
        self,
        name: str,
        environment_suffix: str,
        memory_size: int,
        timeout: int,
        transactions_table_name: Output[str],
        transactions_table_arn: Output[str],
        sessions_table_name: Output[str],
        sessions_table_arn: Output[str],
        subnet_ids: List[Output[str]],
        security_group_id: Output[str],
        tags: dict,
        opts: Optional[ResourceOptions] = None
    ):
        super().__init__("custom:lambda:LambdaStack", name, None, opts)

        # IAM role for Lambda
        lambda_role = aws.iam.Role(
            f"lambda-role-{environment_suffix}",
            assume_role_policy=json.dumps({
                "Version": "2012-10-17",
                "Statement": [{
                    "Action": "sts:AssumeRole",
                    "Principal": {"Service": "lambda.amazonaws.com"},
                    "Effect": "Allow",
                }]
            }),
            tags={**tags, "Name": f"lambda-role-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        # Attach basic Lambda execution policy
        aws.iam.RolePolicyAttachment(
            f"lambda-basic-execution-{environment_suffix}",
            role=lambda_role.name,
            policy_arn="arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
            opts=ResourceOptions(parent=self)
        )

        # Attach VPC access policy
        aws.iam.RolePolicyAttachment(
            f"lambda-vpc-execution-{environment_suffix}",
            role=lambda_role.name,
            policy_arn="arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
            opts=ResourceOptions(parent=self)
        )

        # DynamoDB access policy
        dynamodb_policy = aws.iam.Policy(
            f"lambda-dynamodb-policy-{environment_suffix}",
            policy=Output.all(transactions_table_arn, sessions_table_arn).apply(
                lambda arns: json.dumps({
                    "Version": "2012-10-17",
                    "Statement": [{
                        "Effect": "Allow",
                        "Action": [
                            "dynamodb:GetItem",
                            "dynamodb:PutItem",
                            "dynamodb:UpdateItem",
                            "dynamodb:Query",
                            "dynamodb:Scan"
                        ],
                        "Resource": arns
                    }]
                })
            ),
            tags={**tags, "Name": f"lambda-dynamodb-policy-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        aws.iam.RolePolicyAttachment(
            f"lambda-dynamodb-policy-attachment-{environment_suffix}",
            role=lambda_role.name,
            policy_arn=dynamodb_policy.arn,
            opts=ResourceOptions(parent=self)
        )

        # Payment processor Lambda
        self.payment_processor = aws.lambda_.Function(
            f"payment-processor-{environment_suffix}",
            name=f"payment-processor-{environment_suffix}",
            runtime=aws.lambda_.Runtime.PYTHON3D11,
            handler="payment_processor.lambda_handler",
            role=lambda_role.arn,
            code=pulumi.FileArchive("./lib/lambda"),
            memory_size=memory_size,
            timeout=timeout,
            environment=aws.lambda_.FunctionEnvironmentArgs(
                variables={
                    "TRANSACTIONS_TABLE": transactions_table_name,
                    "ENVIRONMENT": environment_suffix,
                }
            ),
            vpc_config=aws.lambda_.FunctionVpcConfigArgs(
                subnet_ids=subnet_ids,
                security_group_ids=[security_group_id],
            ),
            tags={**tags, "Name": f"payment-processor-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        # Session manager Lambda
        self.session_manager = aws.lambda_.Function(
            f"session-manager-{environment_suffix}",
            name=f"session-manager-{environment_suffix}",
            runtime=aws.lambda_.Runtime.PYTHON3D11,
            handler="session_manager.lambda_handler",
            role=lambda_role.arn,
            code=pulumi.FileArchive("./lib/lambda"),
            memory_size=memory_size,
            timeout=timeout,
            environment=aws.lambda_.FunctionEnvironmentArgs(
                variables={
                    "SESSIONS_TABLE": sessions_table_name,
                    "ENVIRONMENT": environment_suffix,
                }
            ),
            vpc_config=aws.lambda_.FunctionVpcConfigArgs(
                subnet_ids=subnet_ids,
                security_group_ids=[security_group_id],
            ),
            tags={**tags, "Name": f"session-manager-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        # CloudWatch log groups
        log_retention = self._get_log_retention_days(environment_suffix)

        aws.cloudwatch.LogGroup(
            f"payment-processor-logs-{environment_suffix}",
            name=f"/aws/lambda/payment-processor-{environment_suffix}",
            retention_in_days=log_retention,
            tags={**tags, "Name": f"payment-processor-logs-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        aws.cloudwatch.LogGroup(
            f"session-manager-logs-{environment_suffix}",
            name=f"/aws/lambda/session-manager-{environment_suffix}",
            retention_in_days=log_retention,
            tags={**tags, "Name": f"session-manager-logs-{environment_suffix}"},
            opts=ResourceOptions(parent=self)
        )

        self.register_outputs({
            "payment_processor_arn": self.payment_processor.arn,
            "payment_processor_invoke_arn": self.payment_processor.invoke_arn,
            "session_manager_arn": self.session_manager.arn,
            "session_manager_invoke_arn": self.session_manager.invoke_arn,
        })

    def _get_log_retention_days(self, environment: str) -> int:
        """Get log retention days based on environment."""
        retention_map = {
            "dev": 7,
            "staging": 30,
            "prod": 90
        }
        return retention_map.get(environment, 7)

File: archive/pulumi-ts/Pr7324/lib/lambda-stack.ts
/**
 * lambda-stack.ts
 *
 * Defines Lambda function for data processing in VPC.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface LambdaStackArgs {
  region: string;
  vpcId: pulumi.Input<string>;
  privateSubnetIds: pulumi.Input<string>[];
  securityGroupId: pulumi.Input<string>;
  auroraEndpoint: pulumi.Input<string>;
  dynamoDbTableName: pulumi.Input<string>;
  environmentSuffix?: string;
  tags?: pulumi.Input<{ [key: string]: string }>;
}

export class LambdaStack extends pulumi.ComponentResource {
  public readonly function: aws.lambda.Function;
  public readonly role: aws.iam.Role;
  public readonly logGroup: aws.cloudwatch.LogGroup;
  public readonly functionArn: pulumi.Output<string>;
  public readonly functionName: pulumi.Output<string>;

  constructor(
    name: string,
    args: LambdaStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:lambda:LambdaStack', name, args, opts);

    const region = args.region;
    const envSuffix = args.environmentSuffix || 'dev';
    const tags = args.tags || {};

    // Create IAM role for Lambda
    this.role = new aws.iam.Role(
      `${name}-lambda-role`,
      {
        assumeRolePolicy: aws.iam.assumeRolePolicyForPrincipal({
          Service: 'lambda.amazonaws.com',
        }),
        tags: {
          ...tags,
          Name: `${name}-lambda-role-${envSuffix}-e7`,
          Region: region,
        },
      },
      { parent: this }
    );

    // Attach basic Lambda execution policy
    new aws.iam.RolePolicyAttachment(
      `${name}-lambda-basic`,
      {
        role: this.role.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
      },
      { parent: this }
    );

    // Attach VPC execution policy
    new aws.iam.RolePolicyAttachment(
      `${name}-lambda-vpc`,
      {
        role: this.role.name,
        policyArn:
          'arn:aws:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole',
      },
      { parent: this }
    );

    // Create inline policy for DynamoDB access
    new aws.iam.RolePolicy(
      `${name}-lambda-dynamodb-policy`,
      {
        role: this.role.id,
        policy: pulumi.jsonStringify({
          Version: '2012-10-17',
          Statement: [
            {
              Effect: 'Allow',
              Action: [
                'dynamodb:PutItem',
                'dynamodb:GetItem',
                'dynamodb:UpdateItem',
                'dynamodb:Query',
                'dynamodb:Scan',
              ],
              Resource: '*',
            },
          ],
        }),
      },
      { parent: this }
    );

    // Create CloudWatch Log Group
    this.logGroup = new aws.cloudwatch.LogGroup(
      `${name}-lambda-logs`,
      {
        name: `/aws/lambda/${name}-data-processor-${envSuffix}-e7`,
        retentionInDays: 7,
        tags: {
          ...tags,
          Region: region,
        },
      },
      { parent: this }
    );

    // Create Lambda function
    const lambdaCode = `
exports.handler = async (event) => {
  console.log('Event received:', JSON.stringify(event, null, 2));

  const dbEndpoint = process.env.DB_ENDPOINT;
  const tableName = process.env.DYNAMODB_TABLE;
  const region = process.env.AWS_REGION_NAME;

  console.log('Configuration:', { dbEndpoint, tableName, region });

  // In production, this would process data from Aurora and DynamoDB
  const response = {
    statusCode: 200,
    body: JSON.stringify({
      message: 'Data processing completed',
      region: region,
      timestamp: new Date().toISOString(),
    }),
  };

  console.log('Processing complete');
  return response;
};
`;

    this.function = new aws.lambda.Function(
      `${name}-data-processor`,
      {
        name: `${name}-data-processor-${envSuffix}-e7`,
        runtime: 'nodejs20.x',
        handler: 'index.handler',
        role: this.role.arn,
        code: new pulumi.asset.AssetArchive({
          'index.js': new pulumi.asset.StringAsset(lambdaCode),
        }),
        memorySize: 512,
        timeout: 30,
        environment: {
          variables: {
            DB_ENDPOINT: args.auroraEndpoint,
            DYNAMODB_TABLE: args.dynamoDbTableName,
            AWS_REGION_NAME: region,
          },
        },
        vpcConfig: {
          subnetIds: args.privateSubnetIds,
          securityGroupIds: [args.securityGroupId],
        },
        tags: {
          ...tags,
          Name: `${name}-lambda-${envSuffix}-e7`,
          Region: region,
          Purpose: 'data-processing',
        },
      },
      { parent: this, dependsOn: [this.logGroup] }
    );

    this.functionArn = this.function.arn;
    this.functionName = this.function.name;

    this.registerOutputs({
      functionArn: this.function.arn,
      functionName: this.function.name,
      logGroupName: this.logGroup.name,
    });
  }
}
